wdl-lint/src/rules/blanks_between_elements.rs

```diff
@@ -0,0 +1,546 @@
+use wdl_ast::SyntaxNode;
+use wdl_ast::SyntaxToken;
+use wdl_ast::ToSpan;
+use wdl_ast::VisitReason;
+use wdl_ast::Visitor;
+
+use crate::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// The identifier for the blanks between elements rule.
+const ID: &str = "BlanksBetweenElements";
+
+/// Creates an excessive blank line diagnostic.
+fn excess_blank_line(span: Span) -> Diagnostic {
+    Diagnostic::note("extra blank line(s) found")
+        .with_rule(ID)
+        .with_highlight(span)
+        .with_fix("remove blank line(s)")
+}
+
+/// Creates a missing blank line diagnostic.
+fn missing_blank_line(span: Span) -> Diagnostic {
+    Diagnostic::note("missing blank line")
+        .with_rule(ID)
+        .with_highlight(span)
+        .with_fix("add blank line before this element")
+}
+
+/// Track the position within a document
+#[derive(Debug, Default, Clone, Copy, PartialEq)]
+enum State {
+    /// Outside tracked sections
+    #[default]
+    Outside,
+    /// In a `Input` section
+    InputSection,
+    /// In a `Output` section
+    OutputSection,
+    /// In a `Meta` section
+    MetaSection,
+    /// In a `Parameter Meta` section
+    ParameterMetaSection,
+    /// In a `Runtime` section
+    RuntimeSection,
+}
+
+/// Detects unsorted input declarations.
+#[derive(Default, Debug, Clone, Copy)]
+pub struct BlanksBetweenElementsRule {
+    /// Store whether we are in certain blocks
+    state: State,
+}
+
+impl Rule for BlanksBetweenElementsRule {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Ensures that WDL elements are spaced appropriately."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "There should be a blank line between each WDL element at the root indentation level (such \
+         as the import block and any task/workflow definitions) and between sections of a WDL task \
+         or workflow. Never have a blank line when indentation levels are changing (such as \
+         between the opening of a workflow definition and the meta section). There should also \
+         never be blanks within a meta, parameter meta, input, output, or runtime section. See \
+         example for a complete WDL document with proper spacing between elements. Note the blank \
+         lines between meta, parameter meta, input, the first call or first private declaration, \
+         output, and runtime for the example task. The blank line between the workflow definition \
+         and the task definition is also important."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Spacing])
+    }
+}
+
+impl Visitor for BlanksBetweenElementsRule {
+    type State = Diagnostics;
+
+    fn document(
+        &mut self,
+        _: &mut Self::State,
+        reason: VisitReason,
+        _: &Document,
+        _: SupportedVersion,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        if reason == VisitReason::Enter {
```

Comment:
Is this required? The only other `VisitReason` is `VisitReason::Enter`, so this should be redundant.